source/common/quic/envoy_quic_session_cache.cc

```diff
@@ -0,0 +1,168 @@
+  if (state == other) {
+    return true;
+  }
+  if ((state != nullptr && other == nullptr) || (state == nullptr && other != nullptr)) {
+    return false;
+  }
+  return (*state == *other);
+}
+
+} // namespace
+
+EnvoyQuicSessionCache::EnvoyQuicSessionCache(TimeSource& time_source) : time_source_(time_source) {}
+
+EnvoyQuicSessionCache::~EnvoyQuicSessionCache() = default;
+
+void EnvoyQuicSessionCache::Insert(const quic::QuicServerId& server_id,
+                                   bssl::UniquePtr<SSL_SESSION> session,
+                                   const quic::TransportParameters& params,
+                                   const quic::ApplicationState* application_state) {
+  auto it = cache_.find(server_id);
+  if (it == cache_.end()) {
+    // New server ID, add a new entry.
+    createAndInsertEntry(server_id, std::move(session), params, application_state);
+    return;
+  }
+  Entry& entry = it->second;
+  if (params == *entry.params &&
+      doApplicationStatesMatch(application_state, entry.application_state.get())) {
+    // The states are both the same, so we only need to insert the session.
+    entry.pushSession(std::move(session));
+    return;
+  }
+  // States are different, replace the entry.
+  cache_.erase(it);
+  createAndInsertEntry(server_id, std::move(session), params, application_state);
+}
+
+std::unique_ptr<quic::QuicResumptionState>
+EnvoyQuicSessionCache::Lookup(const quic::QuicServerId& server_id, const SSL_CTX* /*ctx*/) {
+  auto it = cache_.find(server_id);
+  if (it == cache_.end()) {
+    return nullptr;
+  }
+  Entry& entry = it->second;
+  const SystemTime system_time = time_source_.systemTime();
+  if (!isSessionValid(entry.peekSession(), system_time)) {
+    cache_.erase(it);
+    return nullptr;
+  }
+  auto state = std::make_unique<quic::QuicResumptionState>();
+  state->tls_session = entry.popSession();
+  if (entry.params != nullptr) {
+    state->transport_params = std::make_unique<quic::TransportParameters>(*entry.params);
+  }
+  if (entry.application_state != nullptr) {
+    state->application_state = std::make_unique<quic::ApplicationState>(*entry.application_state);
+  }
+  return state;
+}
+
+void EnvoyQuicSessionCache::ClearEarlyData(const quic::QuicServerId& server_id) {
+  auto it = cache_.find(server_id);
+  if (it == cache_.end()) {
+    return;
+  }
+  for (bssl::UniquePtr<SSL_SESSION>& session : it->second.sessions) {
+    if (session != nullptr) {
+      session.reset(SSL_SESSION_copy_without_early_data(session.get()));
+    }
+  }
+}
+
+void EnvoyQuicSessionCache::prune() {
+  quic::QuicServerId oldest_id;
+  uint64_t oldest_expiration = std::numeric_limits<uint64_t>::max();
+  const SystemTime system_time = time_source_.systemTime();
+  auto it = cache_.begin();
+  while (it != cache_.end()) {
+    Entry& entry = it->second;
+    SSL_SESSION* session = entry.peekSession();
+    if (!isSessionValid(session, system_time)) {
+      it = cache_.erase(it);
```

Comment:
Is it safe to erase the entry while iterating on the map?

Reply:
Yes. It invalidates the iterator but we don't use it afterwards